### PR TITLE
Parser as lib

### PR DESCRIPTION
**Pull request by jkuebart**
Note: I only created this PR in order to be able to comment the changes.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -6,142 +6,96 @@
...
-	result = luaL_loadfile(L.get(), args[1].c_str()) || lua_pcall(L.get(), 0, 0, 0);
-	if (result){
-		cerr << "error loading \"" << args[1] << "\": " << lua_tostring(L.get(), -1) << endl;
-		return EXIT_FAILURE;
-	}
+		const lua::value subnodes = lua::newtable();
+		for (size_t i = 0; i != sv.size(); ++i){
+			subnodes[1 + i] = sv[i].get<lua::value>();
+		}
+		params["subnodes"] = subnodes;
```
Comment (jkuebart) [inline src/main.cpp]: This is called `values` in the Lua code below.

File: test/adder.lua [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+function rule(r)
+	grammar[1 + #grammar] = r
+end
+
+rule("Add <- (Num '+')* Num")
+rule("Num <- [0-9]+")
+
+actions["Add"] = function(params)
+	res = 0
+	for term in params.values do
```
Comment (jkuebart) [inline test/adder.lua]: This doesn't look like valid Lua. It should be
```lua
local res = 0
for i = 1, #params.values do
    term = params.values[i]
    res = res + term
end
return res
```

File: src/main.cpp [hunk 3]
```diff
@@ -6,142 +6,96 @@
...
-		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
+		(*pegParser)[rule.c_str()].enter = [rule](const char* s, size_t n, any& dt) {
 			auto& indent = *dt.get<int*>();
-			cout << repeat("|  ", indent) << r << " => \"" << shorten(s, n, DEBUG_STRLEN) << "\"?" << endl;
+			cout << repeat("|  ", indent) << rule << " => \"" << shorten(s, n, DEBUG_STRLEN) << "\"?" << endl;
 			indent++;
 		};
 
-		parser[r.c_str()].leave = [r](const char* s, size_t, size_t matchlen, any& value, any& dt) {
+		(*pegParser)[rule.c_str()].leave = [rule, &L](const char* s, size_t n, size_t matchlen, any& value, any& dt) {
```
Comment (jkuebart) [inline src/main.cpp]: As above, `&L` here refers to a variable that does not exist anymore when this lambda is called. But `L` is also unused.

File: src/main.cpp [hunk 4]
```diff
@@ -6,142 +6,96 @@
...
+	// funcName = the name of the associated action in lua's globals registry
 
-	// init lua
-	// Since the parser state holds Lua values, it's important that the Lua VM and lua::scope
-	// remain until after the parser is destroyed. Using a unique_ptr ensures that Lua outlives
-	// any variables created later in this function.
-	std::unique_ptr<lua_State, void(&)(lua_State*)> L(luaL_newstate(), lua_close);
-	lua::scope luascope(L.get());
-	luaL_openlibs(L.get());
+	(*pegParser)[rule.c_str()] = [&L, rule, funcName](const SemanticValues& sv, any& dt){
```
[the remark this answers is not in the document]
Comment (mqnc) [reply src/main.cpp]: That might explain things.

File: src/main.cpp [hunk 5]
```diff
@@ -6,142 +6,96 @@
...
-{
 
-	// check arguments
-	if(args.size()<3){
-		cout << "usage: " << args[0] << " parser.lua file" << endl
-			<< "\tRead the grammar and reduction rules from parser.lua and" << endl
-			<< "\tparse the given file." << endl;
-		return EXIT_FAILURE;
-	}
+void registerReductionRule(parser *pegParser, lua_State *L, const string& rule, const string& funcName){
```
[the remark this answers is not in the document]
Comment (jkuebart) [reply src/main.cpp]: Passing by pointer or reference is independent of being »a heap thing«. Just call `registerReductionRule(*pegParser, …)`.

File: src/main.cpp [hunk 6]
```diff
@@ -6,142 +6,96 @@
...
+	// read action table ( https://stackoverflow.com/questions/6137684/iterate-through-lua-table )
+	lua_pushvalue(L, 2);
+	lua_pushnil(L);
+	while (lua_next(L, -2))
+	{
+		lua_pushvalue(L, -2);
+		string rule = lua_tostring(L, -1); // = key
+		string funcName = ptr2str(lua_topointer(L, -2)); // = value (use the address of the function as its name)
+		lua_pop(L, 1); // pop key
+		lua_setglobal(L, funcName.c_str()); // store function in lua globals (pops from stack)
```
Comment (jkuebart) [inline src/main.cpp]: Note that this does *not* store the function in Lua's registry (as suggested by a comment in `registerReductionRule`) but rather as a global variable which is also visible to Lua code. Storing the value in the registry instead has the advantage that Lua code doesn't see arbitrary global variables appear apparently from nowhere and reduces the risk of inadvertently changing them.

**Comment by jkuebart** [comment]
An alternative implementation for `makeParser()` could be something like this:
```c++
        // Read default action.
        lua_pushvalue(L, 3); // copy the function handle to top
        const lua::value defaultReduce;

        // Read actions table.
        lua_pushvalue(L, 2);
        const lua::value actions;

        // Register actions for all PEG definitions.
        const auto rules = pegParser->get_rule_names();
        for (const auto& rule : rules) {
                const lua::value reduce(actions[rule]);
                if (!reduce.isnil()) {
                        registerReductionRule(*pegParser, rule, reduce);
                } else {
                        registerReductionRule(*pegParser, rule, defaultReduce);
                }
        }
```
The following change would need to be made to `registerReductionRule()`:
```c++
void registerReductionRule(parser& pegParser, const string& rule, const lua::value& reduce) {
        pegParser[rule.c_str()] = [rule, reduce](const SemanticValues& sv, any&) {
                /* … */

                // Return result of reduction function.
                return reduce(params);
        };
}
```

File: src/main.cpp (hunk 6 above)
Comment (mqnc) [reply src/main.cpp]: In that comment I was referring to the... glossary where lua's globals are registered. I was not aware there is yet another, more registry-esque registry which is more suitable. Thanks for pointing that out.

Comment (jkuebart) [reply src/main.cpp]: `lua_getglobal()` and `lua_setglobal()` merely modify the table of global variables. This table can also be accessed using `lua_pushglobaltable()`.

In contrast, the pseudo index [`LUA_REGISTRYINDEX`](https://www.lua.org/manual/5.3/manual.html#4.5) refers to a table that is only accessible to the host.

**Pull request closed without merging**